### PR TITLE
Pixel participation

### PR DESCRIPTION
**Pull request by hashkar**
Adding pixel participation fraction and BPX timeline modules to the DQM - PRELIMINARY WORKING VERSION

### CODE REVIEW

**Comment by jlenain** [comment]
Hi @hashkar !
Thank you! Could you first rebate your branch, to accommodate for the recent #78 and #74 ?
For instance, `start_calib.py` was renamed `start_dqm.py`.
Thanks!

**Review by jlenain — CHANGES_REQUESTED** [review]

File: src/nectarchain/dqm/start_dqm.py [hunk 1]
```diff
@@ -4,9 +4,15 @@
 import time
 
 from camera_monitoring import CameraMonitoring
+<<<<<<< HEAD:src/nectarchain/dqm/start_dqm.py
```
[the remark this answers is not in the document]
Comment (hashkar) [reply src/nectarchain/dqm/start_dqm.py]: I think it should be resolved now?

File: src/nectarchain/dqm/pixel_participation.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+    def ConfigureForRun(self, path, Pix, Samp, Reader1):
+        # define number of pixels and samples
+        self.Pix = Pix
+        self.Samp = Samp
+
+        self.counter_evt = 0
+        self.counter_ped = 0
+
+        self.camera = CameraGeometry.from_name("NectarCam-003").transform_to(EngineeringCameraFrame())
```
Comment (jlenain) [inline src/nectarchain/dqm/pixel_participation.py:22]: Please try not to hardcode the camera geometry version, as e.g. done in other classes.

File: src/nectarchain/dqm/pixel_participation.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        self.camera2 = CameraGeometry.from_name("NectarCam-003").transform_to(EngineeringCameraFrame())
+
+        self.cmap = "gnuplot2"
+        self.cmap2 = "gnuplot2"
+
+
+        self.BadPixels_ped = np.zeros(self.Pix)
```
Comment (jlenain) [inline src/nectarchain/dqm/pixel_participation.py:29]: Please declare all class members within `__init__`.

File: src/nectarchain/dqm/pixel_participation.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            ] = fig1
+            full_name = name + "_Camera_BPX_%sGain.png" % gain_c
+            FullPath = FigPath + full_name
+            self.PixelParticipation_Figures_Names_Dict[
+                "CAMERA-BADPIX-PHY-DISPLAY-%s-GAIN" % gain_c
+            ] = FullPath
+            plt.close()
+
+        if self.counter_ped > 0:
+            fig2, self.disp2 = plt.subplots()
```
Comment (jlenain) [inline src/nectarchain/dqm/pixel_participation.py:123]: These lines are almost an exact duplication of lines 100+. Please try to regroup them together.

File: src/nectarchain/dqm/pixel_participation.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+            self.BadPixels += BadPixels1
+        return None
+
+    def FinishRun(self):
+        self.BadPixels_ped = np.array(self.BadPixels_ped)
+        self.BadPixels = np.array(self.BadPixels)
+
+
+
+    def GetResults(self):
```
Comment (jlenain) [inline src/nectarchain/dqm/pixel_participation.py:59]: Not PEP-8 compliant. Did you enable the `pre-commit` hooks before your commit ?

File: src/nectarchain/dqm/pixel_participation.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+from ctapipe.visualization import CameraDisplay
+from ctapipe.instrument import CameraGeometry
+from ctapipe.coordinates import EngineeringCameraFrame
+import numpy as np
+
+
+class PixelParticipation_HighLowGain(DQMSummary):
+    def __init__(self, gaink):
+        self.k = gaink
+        return None
```
Comment (jlenain) [inline src/nectarchain/dqm/pixel_participation.py:12]: `__init__` should not return anything.

File: src/nectarchain/dqm/pixel_timeline.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+from ctapipe.visualization import CameraDisplay
+from ctapipe.instrument import CameraGeometry
+from ctapipe.coordinates import EngineeringCameraFrame
+import numpy as np
+
+
+class PixelTimeline_HighLowGain(DQMSummary):
+    def __init__(self, gaink):
+        self.k = gaink
+        return None
```
Comment (jlenain) [inline src/nectarchain/dqm/pixel_timeline.py:12]: `__init__` should not return anything.

File: src/nectarchain/dqm/pixel_timeline.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
+from dqm_summary_processor import DQMSummary
+from matplotlib import pyplot as plt
+from ctapipe.visualization import CameraDisplay
+from ctapipe.instrument import CameraGeometry
+from ctapipe.coordinates import EngineeringCameraFrame
+import numpy as np
+
+
+class PixelTimeline_HighLowGain(DQMSummary):
```
Comment (jlenain) [inline src/nectarchain/dqm/pixel_timeline.py:9]: Class names should use CapWords convention

File: src/nectarchain/dqm/pixel_timeline.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+    def __init__(self, gaink):
+        self.k = gaink
+        return None
+
+    def ConfigureForRun(self, path, Pix, Samp, Reader1):
+        # define number of pixels and samples
+        self.Pix = Pix
```
Comment (jlenain) [inline src/nectarchain/dqm/pixel_timeline.py:16]: Duplicated code fragment (12 lines long)

Comment (jlenain) [inline src/nectarchain/dqm/pixel_timeline.py:16]: Please try to define all class members within `__init__`.

File: src/nectarchain/dqm/pixel_timeline.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+        self.Samp = Samp
+
+
+        self.counter_evt = 0
```
Comment (jlenain) [inline src/nectarchain/dqm/pixel_timeline.py:20]: Not PEP-8 compliant. Please enable `pre-commit` hooks before committing.

File: src/nectarchain/dqm/pixel_timeline.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+        self.counter_ped = 0
+
+        self.camera = CameraGeometry.from_name("NectarCam-003").transform_to(EngineeringCameraFrame())
```
Comment (jlenain) [inline src/nectarchain/dqm/pixel_timeline.py:23]: Here also, try to use the camera geometry version coming with the `reader`.

**Review by jlenain — CHANGES_REQUESTED** [review]
Hi @hashkar ,
Thanks for the changes. There are a couple of coding conventions should be adhered to, before merging this PR.

File: src/nectarchain/dqm/pixel_participation.py (hunk 3 above)
Comment (hashkar) [reply src/nectarchain/dqm/pixel_participation.py:29]: Is it ok if I fix the issues related to pixel_participation and pixel_timeline in this PR and then I do another PR to work on the format of the code.

File: src/nectarchain/dqm/pixel_participation.py (hunk 4 above)
Comment (hashkar) [reply src/nectarchain/dqm/pixel_participation.py:123]: this also applies for all the classes in dqm

File: src/nectarchain/dqm/pixel_participation.py (hunk 6 above)
Comment (hashkar) [reply src/nectarchain/dqm/pixel_participation.py:12]: this also applies for the entire dqm as there is a return None in every __init__

File: src/nectarchain/dqm/pixel_participation.py (hunk 3 above)
Comment (jlenain) [reply src/nectarchain/dqm/pixel_participation.py:29]: There is no problem in defining the members in `ConfigureForRun`, I was refereeing to their declaration.

File: src/nectarchain/dqm/pixel_timeline.py (hunk 9 above)
Comment (jlenain) [reply src/nectarchain/dqm/pixel_timeline.py:16]: Better would have to have a common `ConfigureForRun` method in a base class, which would be inherited and called within `pixel_timeline.py` and `pixel_participation.py`.

**Comment by hashkar** [comment]
@jlenain 9 hours for 1 run seems way too much. When I run the DQM on my local machine I can process a run (single file) of ~8000 events in a few minutes with the plots. I usually run it on 3 or 4 different runs before doing a commit.  It takes less time if the plots are not computed. There is no doubt that the DQM needs optimization but are we sure that the  9 hours are due to the DQM code? if I am not mistaken your run has ~20 000 events which is not that much more.  Maybe the plots are not well handled on a server? I know that the charge spectrum plots in charge_integration.py take a while to compute but still, we are talking a few minutes here.

File: src/nectarchain/dqm/pixel_participation.py (hunk 2 above)
Comment (hashkar) [reply src/nectarchain/dqm/pixel_participation.py]: Done

File: src/nectarchain/dqm/pixel_participation.py (hunk 3 above)
Comment (hashkar) [reply src/nectarchain/dqm/pixel_participation.py]: Done

File: src/nectarchain/dqm/pixel_participation.py (hunk 4 above)
Comment (hashkar) [reply src/nectarchain/dqm/pixel_participation.py]: Done

File: src/nectarchain/dqm/pixel_participation.py (hunk 5 above)
Comment (hashkar) [reply src/nectarchain/dqm/pixel_participation.py:74]: enabled now

File: src/nectarchain/dqm/pixel_participation.py (hunk 6 above)
Comment (hashkar) [reply src/nectarchain/dqm/pixel_participation.py]: done

File: src/nectarchain/dqm/pixel_timeline.py (hunk 8 above)
Comment (hashkar) [reply src/nectarchain/dqm/pixel_timeline.py]: done

File: src/nectarchain/dqm/pixel_timeline.py (hunk 9 above)
Comment (hashkar) [reply src/nectarchain/dqm/pixel_timeline.py:29]: done

File: src/nectarchain/dqm/pixel_timeline.py (hunk 10 above)
Comment (hashkar) [reply src/nectarchain/dqm/pixel_timeline.py:31]: done

File: src/nectarchain/dqm/pixel_timeline.py (hunk 11 above)
Comment (hashkar) [reply src/nectarchain/dqm/pixel_timeline.py]: done

**Comment by hashkar** [comment]
@jlenain this PR is ready for review now

File: src/nectarchain/dqm/pixel_timeline.py (hunk 10 above)
Comment (jlenain) [reply src/nectarchain/dqm/pixel_timeline.py:28]: Well, actually I had to run another pass on `pixel_timeline.py` to lint it.

**Review by jlenain — APPROVED** [review]

**Comment by jlenain** [comment]
Support for python 3.8 in `protozfits` has been dropped 2 days ago (see https://gitlab.cta-observatory.org/cta-computing/common/protozfits-python/-/commit/bf475ba48ecb3a443023ef667f409d588af17fc6). Merging anyway.

**Pull request merged**